DepotDownloader/Program.cs

```diff
@@ -9,58 +15,171 @@
+                new Option<bool>(new[] { "--validate", "--verify-all" }, "Include checksum verification of files already downloaded"),
+                new Option<bool>("--manifest-only", "Downloads a human readable manifest for any depots that would be downloaded"),
+
+                new Option<int?>("--cellid", "The overridden CellID of the content server to download from"),
+                new Option<int>("--max-servers", () => 20, "Maximum number of content servers to use"),
+                new Option<int>("--max-downloads", () => 8, "Maximum number of chunks to download concurrently"),
+                new Option<uint?>("--loginid", "A unique 32-bit integer Steam LogonID in decimal, required if running multiple instances of DepotDownloader concurrently"),
+            };
 
-        static async Task<int> MainAsync(string[] args)
+            rootCommand.Handler = CommandHandler.Create<InputModel>(DownloadAsync);
+
+            return new CommandLineBuilder(rootCommand)
+                .UseDefaults()
+                .UseHelpBuilder(ctx => new CustomHelpBuilder(ctx.Console))
+                .Build().InvokeAsync(args);
+        }
+
+        private sealed class CustomHelpBuilder : HelpBuilder
         {
-            if (args.Length == 0)
+            public CustomHelpBuilder(IConsole console) : base(console)
             {
-                PrintUsage();
-                return 1;
             }
 
-            DebugLog.Enabled = false;
+            protected override void AddUsage(ICommand command)
+            {
+                if (command is not RootCommand)
+                    return;
+
+                Console.Out.WriteLine(@$"Examples:
+  - downloading one or all depots for an app:
+    {command.Name} --app <id> [--depot <id> [--manifest <id>]] [--username <username> [--password <password>]]
 
-            AccountSettingsStore.LoadFromFile("account.config");
+  - downloading a workshop item using pubfile id:
+    {command.Name} --app <id> --pubfile <id> [--username <username> [--password <password>]]
 
-            #region Common Options
+  - downloading a workshop item using ugc id:
+    {command.Name} --app <id> --ugc <id> [--username <username> [--password <password>]]"
+                );
+
+                Console.Out.WriteLine();
+            }
+        }
 
-            if (HasParameter(args, "-debug"))
+        public class InputModel
+        {
+            public InputModel(bool debug, uint app, uint[] depot, ulong[] manifest, ulong? ugc, ulong? pubfile, string? branch, string? branchPassword, string[] os, string[] arch, string[] language, bool lowViolence, string? username, string? password, bool rememberPassword, string directory, FileInfo? fileList, bool validate, bool manifestOnly, int? cellId, int maxServers, int maxDownloads, uint? loginId)
```

Comment:
This is pretty ass, does System.CommandLine not have better ways of handling this?

Reply:
I agree. I very much prefer the way that [CommandLineParser](https://github.com/commandlineparser/commandline) handles it.

Reply:
Constructor is not really needed (you can just use `{ get; set; }`) and it will be set with reflection). Its just my obsession with nullability and minimization of reflection.
You don't even need a class containing all the arguments, you can just put it all as arguments in the `DownloadAsync` method

Reply:
> This is pretty ass, does System.CommandLine not have better ways of handling this?

Just realized that yes, there is!
![image](https://user-images.githubusercontent.com/35262707/147292767-f9e42e9b-8dc7-4680-bba6-c12345fa6c42.png)

But because of `https://github.com/dotnet/command-line-api/issues/1550`, currently this very ugly hack https://github.com/js6pak/DepotDownloader/blob/d6d3513629a1e65afe852b0bbe8dd9e9e0169f76/DepotDownloader/Program.cs#L64-L71 is needed

> I agree. I very much prefer the way that [CommandLineParser](https://github.com/commandlineparser/commandline) handles it.

I've played a bit with `CommandLineParser` and `Spectre.Console.Cli` but neither are really flexible enough to replicate the current CLI options.